bbq/compiler/peephole_pass.go

```diff
@@ -0,0 +1,202 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package compiler
+
+import (
+	"sort"
+
+	"github.com/onflow/cadence/bbq/opcode"
+	"github.com/onflow/cadence/interpreter"
+)
+
+type BytecodeShiftPair struct {
+	Offset int
+	Shift  int
+}
+
+type PeepholeOptimizer[E, T any] interface {
+	Optimize(code []E) []E
+}
+
+type PeepholeInstructionOptimizer[E, T any] struct {
+	PatternsByOpcode map[opcode.Opcode][]PeepholePattern
+	compiler         *Compiler[opcode.Instruction, interpreter.StaticType]
+	// bytecodeShifts is a list of where bytecode has been changed
+	// and how much the offset has changed. this is used to patch jumps
+	// after the peephole optimizations have been applied
+	bytecodeShifts []BytecodeShiftPair
+
+	// jumps is a list of offsets in the optimized bytecode that need to be patched
+	jumps []int
+}
+
+func NewPeepholeOptimizer[E, T any](compiler *Compiler[E, T]) PeepholeOptimizer[E, T] {
+	// this map is meant to improve pattern matching performance
+	patternsByOpcode := make(map[opcode.Opcode][]PeepholePattern)
+	for _, pattern := range AllPatterns {
+		if patternsByOpcode[pattern.Opcodes[0]] == nil {
+			patternsByOpcode[pattern.Opcodes[0]] = make([]PeepholePattern, 0)
+		}
+		patternsByOpcode[pattern.Opcodes[0]] = append(patternsByOpcode[pattern.Opcodes[0]], pattern)
+	}
+	// restrict the compiler to opcode.Instruction and static type
+	// so patterns can ignore generics
+	if c, ok := any(compiler).(*Compiler[opcode.Instruction, interpreter.StaticType]); ok {
+		return &PeepholeInstructionOptimizer[E, T]{
+			PatternsByOpcode: patternsByOpcode,
+			compiler:         c,
+		}
+	}
+	return &PeepholeNoopOptimizer[E, T]{}
+}
+
+type PeepholeNoopOptimizer[E, T any] struct {
+}
+
+func (o *PeepholeNoopOptimizer[E, T]) Optimize(code []E) []E {
+	return code
+}
+
+// logic for patching jumps
+// jumps are in order of their target offset (NOT their position in the optimized bytecode),
+// bytecode shifts are also in order of their offset,
+// so we can keep a counter for cumulative shift and patch jumps in the target order
+// ASSUMPTION: peephole patterns should never match a jump instruction
+func (o *PeepholeInstructionOptimizer[E, T]) patchJumps(optimized []opcode.Instruction) {
+	cumShift := 0
+	currentShiftIndex := 0
+	for _, jump := range o.jumps {
+		var jumpTarget int
+		switch ins := optimized[jump].(type) {
+		case opcode.InstructionJump:
+			jumpTarget = int(ins.Target)
+		case opcode.InstructionJumpIfFalse:
+			jumpTarget = int(ins.Target)
+		case opcode.InstructionJumpIfTrue:
+			jumpTarget = int(ins.Target)
+		case opcode.InstructionJumpIfNil:
+			jumpTarget = int(ins.Target)
+		}
```

Comment:
Maybe extract this into a function in the `opcode` package, e.g. `func JumpTarget(instruction opcode.Instruction) (uint16, bool)`. That way we have it centralized in one place